Sources/MuxUploadSDK/InternalUtilities/Reporting/Reporter.swift

```diff
@@ -33,83 +40,248 @@ class Reporter: NSObject {
+        nonStandardInputReasons: [String],
+        options: UploadOptions,
+        standardizationEndTime: Date,
+        standardizationStartTime: Date,
+        uploadCanceled: Bool,
+        uploadURL: URL
+    ) {
+        guard !options.eventTracking.optedOut else {
+            return
+        }
+
+        let data = InputStandardizationFailedEvent.Data(
+            appName: Bundle.main.appName,
+            appVersion: Bundle.main.appVersion,
+            deviceModel: device.model,
+            errorDescription: errorDescription,
+            inputDuration: inputDuration,
+            inputSize: inputSize,
+            maximumResolution: options.inputStandardization.maximumResolution.description,
+            nonStandardInputReasons: nonStandardInputReasons,
+            platformName: device.systemName,
+            platformVersion: device.systemVersion,
+            regionCode: regionCode,
+            sdkVersion: Version.versionString,
+            standardizationStartTime: standardizationStartTime,
+            standardizationEndTime: standardizationEndTime,
+            uploadCanceled: uploadCanceled,
+            uploadURL: uploadURL
+        )
+
+        let event = InputStandardizationFailedEvent(
+            sessionID: sessionID,
+            version: Version.versionString,
+            data: data
+        )
+
+        send(
+            event: event,
+            url: url
+        )
     }
 }
 
 // TODO: Implement as a separate object so the URLSession
 // can become non-optional, which removes a bunch of edge cases
 extension Reporter: URLSessionDelegate, URLSessionTaskDelegate {
     func urlSession(_ session: URLSession, task: URLSessionTask, willPerformHTTPRedirection response: HTTPURLResponse, newRequest request: URLRequest, completionHandler: @escaping (URLRequest?) -> Swift.Void) {
-        if(self.pendingUploadEvent != nil) {
-            if let redirectUrl = request.url, let httpBody = try? serializePendingEvent() {
-                let request = self.generateRequest(
-                    url: redirectUrl,
-                    httpBody: httpBody
-                )
-                completionHandler(request)
+        if let pendingEvent = pendingEvents[ObjectIdentifier(task)], let redirectURL = request.url {
+            guard let httpBody = try? jsonEncoder.encode(pendingEvent) else {
+                completionHandler(nil)
+                return
             }
+
+            // TODO: This can be URLRequest instead of NSMutableURLRequest
+            // test URLRequest-based construction in case
+            // for any weirdness
+            let request = NSMutableURLRequest.makeJSONPost(
+                url: redirectURL,
+                httpBody: httpBody
+            )
+
+            completionHandler(request as URLRequest)
         }
     }
+
+    func urlSession(_ session: URLSession, task: URLSessionTask, didCompleteWithError error: Error?) {
+        pendingEvents[
+            ObjectIdentifier(task)
```

Comment:
We rely on a hash of the `URLSessionTask` pointer ([`ObjectIdentifier`](https://developer.apple.com/documentation/swift/objectidentifier)) to correlate a request with a pending event.